default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+    def _set_stats_to_default(self):
+        """Set stats to default.
+
+        Call this if design is invalid to avoid miscalculation of ratings."""
+        self.structure = 0
+        self.attacks.clear()
+        self.shields = 0
+        self.fuel = 0.0001
+        self.speed = 0.0001
+        self.stealth = 0.0001
+        self.detection = 0.0001
+        self.troops = 0
+        self.colonisation = -1
+        self.production_cost = 9999999
+        self.production_time = 1
+
+    def update_hull(self, hullname):
+        """Set hull of the design.
+
+        :param hullname: string - name of hull"""
+        self.hull = fo.getHullType(hullname)
+
+    def update_parts(self, partname_list):
+        """Set both partnames and parts attributes.
+
+        :param partname_list: list of strings (partnames)"""
+        local__get_part_type = _get_part_type  # local reference to improve lookup time
+        self.partnames = partname_list
+        self.parts = [local__get_part_type(part) for part in partname_list if part]
+
+    def update_species(self, speciesname):
+        """Set the piloting species.
+
+        :param speciesname: string - name of species
+        """
+        self.species = speciesname
+
+    def update_stats(self):
+        """Calculate and update all stats of the design.
+
+        Default stats if no hull in design."""
+        if not self.hull:
+            print "WARNING: Tried to update stats of design without hull. Reset values to default."
+            self._set_stats_to_default()
+            return
+
+        local_cost_cache = Cache.production_cost[self.pid]
+        local_time_cache = Cache.production_time[self.pid]
+
+        # read out hull stats
+        self.structure = self.hull.structure
+        self.fuel = self.hull.fuel
+        self.speed = self.hull.starlaneSpeed
+        self.stealth = self.hull.stealth
+        self.attacks.clear()
+        self.detection = 0  # ToDo: Add self.hull.detection once available in interface
+        self.shields = 0    # ToDo: Add self.hull.shields if added to interface
+        self.troops = 0     # ToDo: Add self.hull.troops if added to interface
+        self.production_cost = local_cost_cache[self.hull.name]
+        self.production_time = local_time_cache[self.hull.name]
+        self.colonisation = -1  # -1 as 0 corresponds to outpost pod (capacity = 0)
+
+        # read out part stats
+        shield_counter = cloak_counter = detection_counter = colonization_counter = 0  # to deal with Non-stacking parts
+        for part in self.parts:
+            self.production_cost += local_cost_cache[part.name]
+            self.production_time = max(self.production_time, local_time_cache[part.name])
+            pc = part.partClass
+            cap = part.capacity
+            if pc in FUEL:
+                self.fuel += cap
+            elif pc in ENGINES:
+                self.speed += cap
+            elif pc in COLONISATION:
+                colonization_counter += 1
```

Comment:
Stacking should now always choose the strongest part (not sure how colo pods are handled). Anyway, setting the relevant stat to zero always means a worse design rating as the cost increases compared to an empty slot while the baserating will always be smaller or equal than with an empty slot instead.
I could obviously also add the correct stacking behaviour but I think the approach I chose makes the intention more clear.